### PR TITLE
Change order of CORS and authenticatedHandler for secureHandler.

### LINKED ISSUE #16431: apiserver: CORS only works in insecure mode?
I am writing a client-side AngularJS app that uses XMLHttpRequest trying to talk to the apiserver. The apiserver was brought up with the following flags, especially with `cors-allowed-origins` hoping to allow CORS.

```
--cors-allowed-origins=.* 
--insecure-bind-address=127.0.0.1 
--insecure-port=8080 
--bind-address=0.0.0.0 
--secure-port=443
```

and I configured apiserver to support basic auth with https.

However, the pre-flight OPTIONS request failed with 401. Below is my rationale:

- For CORS to work, the client will always first send a pre-flight OPTIONS request to the api endpoint

- When the apiserver runs in the secure mode (with authn), for each request the master will use ["m.Handler"](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/app/server.go#L597) to process the request; whereas an ["m.InsecureHandler"](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/app/server.go#L650) will be used if running in insecure mode

- The m.InsecureHandler will handle CORS [correctly](https://github.com/kubernetes/kubernetes/blob/52136bdb68d7ac34a702f9d962e141664a31ec0f/pkg/master/master.go#L720) without performing any auth.

- In comparison, the m.Handler will add authn functionality *before* CORS, as shown [here](https://github.com/kubernetes/kubernetes/blob/52136bdb68d7ac34a702f9d962e141664a31ec0f/pkg/master/master.go#L731), where the [handlers.NewRequestAuthenticator](https://github.com/kubernetes/kubernetes/blob/52136bdb68d7ac34a702f9d962e141664a31ec0f/pkg/master/master.go#L727) will add auth *before* the InsecureHandler that handles CORS. In short, in the secure mode, we *first* authenticate *all* requests (including OPTIONS) *before* detecting if it is an OPTIONS request

- Meanwhile (unfortunately), as the OPTIONS request cannot carry any auth headers (by w3c standards), the OPTIONS requests will *always* fail the authentication.

Is this expected behavior or did I miss anything? If it is expected, then what is the recommended way to allow CORS with authn on? I am thinking of using a proxy in front of the apiserver

### PR DESCRIPTION
**Pull request by pendoragon**
Per discussion in #16431 . This PR change the handler wrapping order for secureHandler.

Fixes #16431

### CODE REVIEW

**Comment by Pendoragon** [comment]
I signed it!

File: pkg/master/master.go [hunk 1]
```diff
@@ -738,6 +729,17 @@ func (m *Master) init(c *Config) {
 		handler = authenticatedHandler
 	}
 
+	if len(c.CorsAllowedOriginList) > 0 {
```
Comment (lavalamp) [inline pkg/master/master.go]: Please add a comment explaining why this needs to happen before the auth check.

**Comment by Pendoragon** [comment]
comment addressed

**Comment by lavalamp** [comment]
LGTM, thanks!

**Comment by wojtek-t** [comment]
@k8s-bot unit test this please

**Comment by wojtek-t** [comment]
@k8s-bot test this please

**Pull request merged**